### PR TITLE
[Merged by Bors] - feat(Probability/Independence): define independence wrt a kernel and a measure

### PR DESCRIPTION
**Pull request by RemyDegenne**
We introduce a new notion of independence with respect to a kernel and a measure. The plan is to eventually express both independence and conditional independence as particular cases of this new notion (see #6098).

Two sigma-algebras `m` and `m'` are said to be independent with respect to a kernel `κ` and a measure `μ` if for all `m`-measurable sets `t₁` and `m'`-measurable sets `t₂`, `∀ᵐ a ∂μ, κ a (t₁ ∩ t₂) = κ a t₁ * κ a t₂`.

Independence is the special case where `κ` is a constant kernel. Conditional independence can be defined by using the conditional expectation kernel `condexpKernel`. 

---

The file is a copy of Independence/Basic with minor modifications to use the generalized independence definition.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Probability/Independence/Kernel.lean [hunk 1]
```diff
@@ -0,0 +1,916 @@
...
+
+end FromPiSystemsToMeasurableSpaces
+
+section IndepSet
+
+/-! ### Independence of measurable sets
+
+We prove the following equivalences on `IndepSet`, for measurable sets `s, t`.
+* `IndepSetₖ s t μ ↔ μ (s ∩ t) = μ s * μ t`,
+* `IndepSetₖ s t μ ↔ IndepSetsₖ {s} {t} μ`.
```
Comment (sgouezel) [inline Mathlib/Probability/Independence/Kernel.lean:598]: Shouldn't kappa appear in these statements?

File: Mathlib/Probability/Independence/Kernel.lean [hunk 2]
```diff
@@ -0,0 +1,916 @@
...
+  rintro S π _hπ
+  simp_rw [Set.indicator_const_preimage_eq_union]
+  refine' @hs S (fun i => ite (1 ∈ π i) (s i) ∅ ∪ ite ((0 : β) ∈ π i) (s i)ᶜ ∅) fun i _hi => _
+  have hsi : MeasurableSet[generateFrom {s i}] (s i) :=
+    measurableSet_generateFrom (Set.mem_singleton _)
+  refine'
+    MeasurableSet.union (MeasurableSet.ite' (fun _ => hsi) fun _ => _)
+      (MeasurableSet.ite' (fun _ => hsi.compl) fun _ => _)
+  · exact @MeasurableSet.empty _ (generateFrom {s i})
+  · exact @MeasurableSet.empty _ (generateFrom {s i})
```
Comment (sgouezel) [inline Mathlib/Probability/Independence/Kernel.lean:912]: In this file, sometimes you indent things after `classical`, and sometimes you don't. Could you make a choice once and for all? (Personally, I would prefer not indenting, but it's up to you!)

**Review by sgouezel — COMMENTED** [review]
LGTM, thanks!

File: Mathlib/Probability/Independence/Kernel.lean [hunk 3]
```diff
@@ -0,0 +1,916 @@
...
+  change IndepSetsₖ ((fun x => MeasurableSet[m x]) i) ((fun x => MeasurableSet[m x]) j) κ μ
+  exact iIndepSetsₖ.indepSetsₖ h_indep hij
+
+theorem iIndepFunₖ.indepFunₖ {_mΩ : MeasurableSpace Ω}
+    {κ : kernel α Ω} {μ : Measure α} {β : ι → Type _}
+    {m : ∀ x, MeasurableSpace (β x)} {f : ∀ i, Ω → β i} (hf_Indepₖ : iIndepFunₖ m f κ μ) {i j : ι}
+    (hij : i ≠ j) : IndepFunₖ (f i) (f j) κ μ :=
+  hf_Indepₖ.indepₖ hij
+
+end FromIndepToIndep
```
Comment (RemyDegenne) [inline Mathlib/Probability/Independence/Kernel.lean:284]:
```suggestion
end FromiIndepToIndep
```

File: Mathlib/Probability/Independence/Kernel.lean [hunk 4]
```diff
@@ -0,0 +1,916 @@
...
+
+end FromMeasurableSpacesToSetsOfSets
+
+section FromPiSystemsToMeasurableSpaces
+
+/-! ### Independence of generating π-systems implies independence of measurable space structures -/
+
+variable {_mα : MeasurableSpace α}
+
+private theorem IndepSetsₖ.indepₖ_aux {m₂ m : MeasurableSpace Ω}
```
Comment (RemyDegenne) [inline Mathlib/Probability/Independence/Kernel.lean:322]:
```suggestion
theorem IndepSetsₖ.indepₖ_aux {m₂ m : MeasurableSpace Ω}
```

File: Mathlib/Probability/Independence/Kernel.lean [hunk 5]
```diff
@@ -0,0 +1,917 @@
...
+
+variable {_mα : MeasurableSpace α}
+
+/-- A family of sets of sets `π : ι → set (set Ω)` is independent with respect to a kernel `κ` and
+a measure `μ` if for any finite set of indices `s = {i_1, ..., i_n}`, for any sets
+`f i_1 ∈ π i_1, ..., f i_n ∈ π i_n`, then `∀ᵐ a ∂μ, κ a (⋂ i in s, f i) = ∏ i in s, κ a (f i)`.
+It will be used for families of pi_systems. -/
+def iIndepSets {_mΩ : MeasurableSpace Ω}
+    (π : ι → Set (Set Ω)) (κ : kernel α Ω) (μ : Measure α := by volume_tac) :
+    Prop :=
```
Comment (sgouezel) [inline Mathlib/Probability/Independence/Kernel.lean:64]:
```suggestion
    (π : ι → Set (Set Ω)) (κ : kernel α Ω) (μ : Measure α := by volume_tac) : Prop :=
```

File: Mathlib/Probability/Independence/Kernel.lean [hunk 6]
```diff
@@ -0,0 +1,917 @@
...
+`∀ᵐ a ∂μ, κ a (t₁ ∩ t₂) = κ a (t₁) * κ a (t₂)` -/
+def Indep (m₁ m₂ : MeasurableSpace Ω) {_mΩ : MeasurableSpace Ω} (κ : kernel α Ω)
+    (μ : Measure α := by volume_tac) : Prop :=
+  IndepSets {s | MeasurableSet[m₁] s} {s | MeasurableSet[m₂] s} κ μ
+
+/-- A family of sets is independent if the family of measurable space structures they generate is
+independent. For a set `s`, the generated measurable space has measurable sets `∅, s, sᶜ, univ`. -/
+def iIndepSet {_mΩ : MeasurableSpace Ω} (s : ι → Set Ω) (κ : kernel α Ω)
+  (μ : Measure α := by volume_tac) : Prop :=
+iIndep (fun i ↦ generateFrom {s i}) κ μ
```
Comment (sgouezel) [inline Mathlib/Probability/Independence/Kernel.lean:91]:
```suggestion
    (μ : Measure α := by volume_tac) : Prop :=
  iIndep (fun i ↦ generateFrom {s i}) κ μ
```

File: Mathlib/Probability/Independence/Kernel.lean [hunk 7]
```diff
@@ -0,0 +1,917 @@
...
+
+/-- Two sets are independent if the two measurable space structures they generate are independent.
+For a set `s`, the generated measurable space structure has measurable sets `∅, s, sᶜ, univ`. -/
+def IndepSet {_mΩ : MeasurableSpace Ω} (s t : Set Ω) (κ : kernel α Ω)
+  (μ : Measure α := by volume_tac) : Prop :=
+Indep (generateFrom {s}) (generateFrom {t}) κ μ
```
Comment (sgouezel) [inline Mathlib/Probability/Independence/Kernel.lean:97]:
```suggestion
    (μ : Measure α := by volume_tac) : Prop :=
  Indep (generateFrom {s}) (generateFrom {t}) κ μ
```

File: Mathlib/Probability/Independence/Kernel.lean [hunk 8]
```diff
@@ -0,0 +1,917 @@
...
+
+/-- A family of functions defined on the same space `Ω` and taking values in possibly different
+spaces, each with a measurable space structure, is independent if the family of measurable space
+structures they generate on `Ω` is independent. For a function `g` with codomain having measurable
+space structure `m`, the generated measurable space structure is `measurable_space.comap g m`. -/
+def iIndepFun {_mΩ : MeasurableSpace Ω} {β : ι → Type _} (m : ∀ x : ι, MeasurableSpace (β x))
+  (f : ∀ x : ι, Ω → β x) (κ : kernel α Ω)
+  (μ : Measure α := by volume_tac) : Prop :=
+iIndep (fun x ↦ MeasurableSpace.comap (f x) (m x)) κ μ
```
Comment (sgouezel) [inline Mathlib/Probability/Independence/Kernel.lean:106]:
```suggestion
    (f : ∀ x : ι, Ω → β x) (κ : kernel α Ω)
    (μ : Measure α := by volume_tac) : Prop :=
  iIndep (fun x ↦ MeasurableSpace.comap (f x) (m x)) κ μ
```

File: Mathlib/Probability/Independence/Kernel.lean [hunk 9]
```diff
@@ -0,0 +1,917 @@
...
+
+/-- Two functions are independent if the two measurable space structures they generate are
+independent. For a function `f` with codomain having measurable space structure `m`, the generated
+measurable space structure is `MeasurableSpace.comap f m`. -/
+def IndepFun {β γ} {_mΩ : MeasurableSpace Ω} [mβ : MeasurableSpace β] [mγ : MeasurableSpace γ]
+  (f : Ω → β) (g : Ω → γ) (κ : kernel α Ω)
+  (μ : Measure α := by volume_tac) : Prop :=
+Indep (MeasurableSpace.comap f mβ) (MeasurableSpace.comap g mγ) κ μ
```
Comment (sgouezel) [inline Mathlib/Probability/Independence/Kernel.lean:114]:
```suggestion
    (f : Ω → β) (g : Ω → γ) (κ : kernel α Ω)
    (μ : Measure α := by volume_tac) : Prop :=
  Indep (MeasurableSpace.comap f mβ) (MeasurableSpace.comap g mγ) κ μ
```

**Review by sgouezel — COMMENTED** [review]
LGTM, thanks!
bors d+

**Pull request closed without merging**